kubernetes-client/src/main/java/io/fabric8/kubernetes/client/CustomResource.java

```diff
@@ -15,60 +15,92 @@
  */
 package io.fabric8.kubernetes.client;
 
+import java.util.Locale;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
 import com.fasterxml.jackson.databind.JsonDeserializer;
 import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
 import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.Namespaced;
 import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.client.utils.ApiVersionUtil;
+import io.fabric8.kubernetes.client.utils.Pluralize;
+import io.fabric8.kubernetes.client.utils.Utils;
+import io.fabric8.kubernetes.model.annotation.Plural;
 import io.sundr.builder.annotations.Buildable;
 import lombok.ToString;
 
+import static io.fabric8.kubernetes.client.utils.Utils.isNullOrEmpty;
+
 /**
  * A base class for implementing a custom resource kind
  */
 @JsonDeserialize(
-    using = JsonDeserializer.None.class
+  using = JsonDeserializer.None.class
 )
 @ToString
 @Buildable(builderPackage = "io.fabric8.kubernetes.api.builder", editableEnabled = false)
 public abstract class CustomResource implements HasMetadata {
+  public static final String NAMESPACE_SCOPE = "Namespaced";
+  public static final String CLUSTER_SCOPE = "Cluster";
+  private ObjectMeta metadata = new ObjectMeta();
+  
+  @JsonIgnore
+  private String plural;
+  
+  @JsonIgnore
+  private String crdName;
+  
   private String kind;
   private String apiVersion;
-  private ObjectMeta metadata = new ObjectMeta();
-
+  private final String scope;
+  
   public CustomResource() {
-    this.kind = getClass().getSimpleName();
-  }
-
-  public CustomResource(String kind) {
-    this.kind = kind;
+    scope = this instanceof Namespaced ? NAMESPACE_SCOPE : CLUSTER_SCOPE;
   }
-
+  
   @Override
   public String toString() {
     return "CustomResource{" +
-        "kind='" + kind + '\'' +
-        ", apiVersion='" + apiVersion + '\'' +
-        ", metadata=" + metadata +
-        '}';
-  }
-
-  @Override
-  public String getKind() {
-    return kind;
-  }
-
-  public void setKind(String kind) {
-    this.kind = kind;
+      "kind='" + getKind() + '\'' +
+      ", apiVersion='" + getApiVersion() + '\'' +
+      ", metadata=" + metadata +
+      '}';
   }
-
+  
   @Override
   public String getApiVersion() {
+    if (apiVersion == null) {
+      final String apiVersion = HasMetadata.super.getApiVersion();
+      if (isNullOrEmpty(apiVersion)) {
+        throw new IllegalArgumentException(getClass().getName() + " CustomResource must provide an API version using @ApiGroup and @ApiVersion annotations");
+      }
+      this.apiVersion = apiVersion;
```

Comment:
I feel this logic with side-effects in a getter method can bring serious consequences in the future.
Does it make sense to move the "initialization-by-annotation" procedures to the constructor?

Reply:
I originally had most of the code in the constructor but had issues with deserialization where the no-arg constructor would fail… I will double check as I might have missed things.

Reply:
This is exactly the kind of problem we might face. Since this method is invoked upon serialization, it's possible that we end up with non-deterministic behaviors.
The empty-args constructor should be calling the all-args constructor.

Reply:
Putting the code in the constructor doesn't work because when you create a CR from `CustomResourceBuilder`, there's no annotation. Checking for `null` in the getter before doing anything should ensure that if the version has already been set (i.e. in a deserialization context), we don't check the annotation.
The possible solution is to not throw an exception if no annotation is provided… I agree, though, that we should avoid side-effects in getters.